File: src/main/java/com/nurkiewicz/tsclass/parser/ast/expr/AdditiveExpression.java
package com.nurkiewicz.tsclass.parser.ast.expr;

import lombok.EqualsAndHashCode;
import lombok.Value;

@Value
@EqualsAndHashCode(callSuper = true)
public class AdditiveExpression extends BinaryExpression {

    private final Operator operator;

    public AdditiveExpression(Expression left, Operator operator, Expression right) {
        super(left, right);
        this.operator = operator;
    }

    public static AdditiveExpression add(Expression left, Expression right) {
        return new AdditiveExpression(left, Operator.PLUS, right);
    }

    public static AdditiveExpression sub(Expression left, Expression right) {
        return new AdditiveExpression(left, Operator.MINUS, right);
    }

    public String toString() {
        return "(" + getLeft() + " " + operator + " " + getRight() + ")";
    }

    public enum Operator {
        PLUS("+"), MINUS("-");

        private final String s;

        Operator(String s) {
            this.s = s;
        }


        public static Operator of(String s) {
            switch(s) {
                case "+": return PLUS;
                case "-": return MINUS;
                default:
                    throw new IllegalArgumentException(s);
            }
        }
        @Override
        public String toString() {
            return s;
        }

    }

}

File: src/main/kotlin/com/nurkiewicz/tsclass/parser/ast/expr/AdditiveExpression.kt
package com.nurkiewicz.tsclass.parser.ast.expr

data class AdditiveExpression(val left: Expression, val operator: Operator, val right: Expression): Expression {

    override fun toString(): String {
        return "($left $operator $right)"
    }

    enum class Operator private constructor(private val s: String) {
        PLUS("+"), MINUS("-");

        override fun toString(): String {
            return s
        }

        companion object {


            fun of(s: String): Operator {
                when (s) {
                    "+" -> return PLUS
                    "-" -> return MINUS
                    else -> throw IllegalArgumentException(s)
                }
            }
        }

    }

    companion object {

        @JvmStatic
        fun add(left: Expression, right: Expression): AdditiveExpression {
            return AdditiveExpression(left, Operator.PLUS, right)
        }

        @JvmStatic
        fun sub(left: Expression, right: Expression): AdditiveExpression {
            return AdditiveExpression(left, Operator.MINUS, right)
        }
    }

}
